File: Javascript/src/01.AlgorithmicToolbox/Week4DivideAndConquer/inversions.js
/**
 * Problem 4 - Inversions
 * Given a sequence of n integers a0, a1,..., a(n-1), find the number of inversions,
 * where an inversion is a pair of indices i, j where 0 <= i < j < n and ai > aj.
 * @param {Array.<number>} sequence sequence of integers
 * @returns {number} number of inversions
 */
export function inversions(sequence) {
  return inversionsHelper(sequence, 0, sequence.length);
}

function inversionsHelper(sequence, startIndex, endIndex) {
  if (endIndex - startIndex === 1) {
    return 0;
  }
  const middleIndex = startIndex + Math.floor((endIndex - startIndex) / 2);
  const leftInversions = inversionsHelper(sequence, startIndex, middleIndex);
  const rightInversions = inversionsHelper(sequence, middleIndex, endIndex);
  return (
    mergeWithInversions(sequence, startIndex, middleIndex, endIndex) +
    leftInversions +
    rightInversions
  );
}

function mergeWithInversions(sequence, startIndex, middleIndex, endIndex) {
  let inversions = 0;
  const merged = [];
  let leftIndex = startIndex;
  let rightIndex = middleIndex;
  // merge by comparing smallest of each of left, right
  while (leftIndex < middleIndex && rightIndex < endIndex) {
    if (sequence[leftIndex] <= sequence[rightIndex]) {
      merged.push(sequence[leftIndex]);
      ++leftIndex;
    } else {
      merged.push(sequence[rightIndex]);
      // inversions with all remaining left elements
      inversions += middleIndex - leftIndex;
      ++rightIndex;
    }
  }
  // leftover elements
  while (leftIndex < middleIndex) {
    merged.push(sequence[leftIndex]);
    ++leftIndex;
  }
  while (rightIndex < endIndex) {
    merged.push(sequence[rightIndex]);
    ++rightIndex;
  }
  // replace in existing sequence
  for (let i = 0; i < merged.length; ++i) {
    sequence[startIndex + i] = merged[i];
  }
  return inversions;
}

File: Rust/src/_01_algorithmic_toolbox/week4_divide_and_conquer/inversions.rs
///
/// Problem 4 - Inversions
/// Given a sequence of n integers a0, a1,..., a(n-1), returns the number of inversions,
/// where an inversion is a pair of indices i, j where 0 <= i < j < n and ai > aj.
pub fn inversions(sequence: &mut [i64]) -> usize {
    let sequence_length = sequence.len();
    if sequence_length == 1 {
        return 0;
    }
    let middle_index = sequence_length / 2;
    let left_inversions = inversions(&mut sequence[..middle_index]);
    let right_inversions = inversions(&mut sequence[middle_index..]);
    merge_with_inversions(sequence, middle_index) + left_inversions + right_inversions
}

fn merge_with_inversions(sequence: &mut [i64], middle_index: usize) -> usize {
    let mut inversions = 0;
    let mut left_iter = sequence[..middle_index].to_vec().into_iter();
    let mut right_iter = sequence[middle_index..].to_vec().into_iter();
    let mut left_value = left_iter.next();
    let mut right_value = right_iter.next();
    let mut left_count = middle_index;
    for value in sequence {
        match (left_value, right_value) {
            (Some(left), Some(right)) if left <= right => {
                *value = left;
                left_value = left_iter.next();
                left_count -= 1;
            }
            (Some(_left), Some(right)) => {
                *value = right;
                right_value = right_iter.next();
                inversions += left_count;
            }
            (Some(left), None) => {
                *value = left;
                left_value = left_iter.next();
            }
            (None, Some(right)) => {
                *value = right;
                right_value = right_iter.next();
            }
            _ => panic!("unreachable"),
        }
    }
    inversions
}

#[cfg(test)]
mod test {
    use super::*;
    use rand;

    fn is_sorted<I>(data: I) -> bool
    where
        I: IntoIterator,
        I::Item: Ord,
    {
        let mut it = data.into_iter();
        match it.next() {
            None => true,
            Some(first) => it
                .scan(first, |state, next| {
                    let cmp = *state <= next;
                    *state = next;
                    Some(cmp)
                })
                .all(|b| b),
        }
    }

    #[test]
    fn sorts_array_of_random_elements() {
        let mut random_elements: [i64; 32] = rand::random();
        inversions(&mut random_elements);
        assert_eq!(is_sorted(random_elements.iter()), true)
    }

    #[test]
    fn gets_correct_number_of_inversions() {
        assert_eq!(inversions(&mut vec![1, 2, 3]), 0);
        assert_eq!(inversions(&mut vec![3, 1, 2, 3, 1, 2]), 7);
        assert_eq!(inversions(&mut vec![0, -1, 1, -2, 2]), 4);
    }
}
